rllib/examples/curriculum_learning.py

```diff
@@ -0,0 +1,99 @@
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--run", type=str, default="PPO")
+parser.add_argument("--torch", action="store_true")
+parser.add_argument("--as-test", action="store_true")
+parser.add_argument("--stop-iters", type=int, default=50)
+parser.add_argument("--stop-timesteps", type=int, default=200000)
+parser.add_argument("--stop-reward", type=float, default=10000.0)
+
+
+def curriculum_fn(train_results: dict, task_settable_env: TaskSettableEnv,
+                  env_ctx: EnvContext) -> TaskType:
+    """Function returning a possibly new task to set `task_settable_env` to.
+
+    Args:
+        train_results (dict): The train results returned by Trainer.train().
+        task_settable_env (TaskSettableEnv): A single TaskSettableEnv object
+            used inside any worker and at any vector position. Use `env_ctx`
+            to get the worker_index, vector_index, and num_workers.
+        env_ctx (EnvContext): The env context object (i.e. env's config dict
+            plus properties worker_index, vector_index and num_workers) used
+            to setup the `task_settable_env`.
+
+    Returns:
+        TaskType: The task to set the env to. This may be the same as the
+            current one.
+    """
+    # Our env supports tasks 1 (default) to 5.
+    # With each task, rewards get scaled up by a factor of 10, such that:
+    # Level 1: Expect rewards between 0.0 and 1.0.
+    # Level 2: Expect rewards between 1.0 and 10.0, etc..
+    # We will thus raise the level/task each time we hit a new power of 10.0
+    new_task = int(np.log10(train_results["episode_reward_mean"]) + 2.1)
+    # Clamp between valit values, just in case:
+    new_task = max(min(new_task, 5), 1)
+    print(f"Worker #{env_ctx.worker_index} vec-idx={env_ctx.vector_index}"
+          f"\nR={train_results['episode_reward_mean']}"
+          f"\nSeting env to task={new_task}")
+    return new_task
+
+
+if __name__ == "__main__":
+    args = parser.parse_args()
+    ray.init()
+
+    # Can also register the env creator function explicitly with:
+    # register_env(
+    #     "curriculum_env", lambda config: CurriculumCapableEnv(config))
+
+    config = {
+        "env": CurriculumCapableEnv,  # or "curriculum_env" if registered above
+        "env_config": {
+            "start_level": 1,
+        },
+        "num_workers": 2,  # parallelism
+        "num_envs_per_worker": 5,
+        "env_task_fn": curriculum_fn,
+        # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
+        "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
+        "framework": "torch" if args.torch else "tf",
+    }
+
+    stop = {
+        "training_iteration": args.stop_iters,
+        "timesteps_total": args.stop_timesteps,
+        "episode_reward_mean": args.stop_reward,
+    }
+
+    results = tune.run(args.run, config=config, stop=stop, verbose=2)
+
+    if args.as_test:
+        check_learning_achieved(results, args.stop_reward)
+    ray.shutdown()
```

Comment:
you don't need this right?

1. in local case, ray will shutdown when process exits
2. in remote case, ray will disconnect when process exits

Reply:
I remember a time when not calling `shutdown` at the end of tests would lead to re-init errors when we run these tests in e.g. the CI.